ambry-router/src/main/java/com.github.ambry.router/CryptoWorker.java

```diff
@@ -0,0 +1,296 @@
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Thread responsible for encrypting and decrypting blob content and per blob keys. Worker Thread listens to a job queue
+ * for new jobs and processes it one by one.
+ */
+public class CryptoWorker implements Runnable {
+  private final BlockingQueue<CryptoJob> jobQueue;
+  private final CryptoService cryptoService;
+  private final KeyManagementService kms;
+  private final AtomicBoolean enabled;
+  private CountDownLatch shutdownLatch;
+
+  private static final Logger logger = LoggerFactory.getLogger(CryptoWorker.class);
+
+  /**
+   * Instantiates {@link CryptoWorker} with the job queue, {@link CryptoService} and {@link KeyManagementService}
+   * @param jobQueue the {@link BlockingQueue} that contains the jobs
+   * @param cryptoService the {@link CryptoService} to use to encrypt or decrypt
+   * @param kms the {@link KeyManagementService} to fetch keys from
+   */
+  CryptoWorker(BlockingQueue<CryptoJob> jobQueue, CryptoService cryptoService, KeyManagementService kms) {
+    this.jobQueue = jobQueue;
+    this.cryptoService = cryptoService;
+    this.kms = kms;
+    enabled = new AtomicBoolean(false);
+  }
+
+  @Override
+  public void run() {
+    enabled.set(true);
+    shutdownLatch = new CountDownLatch(1);
+    try {
+      while (enabled.get()) {
+        try {
+          CryptoJob cryptoJob = jobQueue.poll(10, TimeUnit.MILLISECONDS);
+          if (cryptoJob != null) {
+            cryptoJob.doOperation(cryptoService, kms);
+          }
+        } catch (InterruptedException e) {
+          logger.error("InterruptedException thrown in CryptoWorker");
+        }
+      }
+    } finally {
+      shutdownLatch.countDown();
+    }
+  }
+
+  /**
+   * Shuts down the thread by closing all pending jobs
+   */
+  void close() {
+    if (enabled.compareAndSet(true, false)) {
+      logger.info("Closing CryptoWorker");
+      closePendingJobs(new GeneralSecurityException("Shutting down EncryptDecrypt Thread"));
+    }
+    if (shutdownLatch != null) {
+      try {
+        shutdownLatch.await(1, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        logger.error("InterruptedException thrown while waiting for CryptoWorker to shutdown");
+      }
+    }
+  }
+
+  /**
+   * Closes all pending jobs with the given {@code e}
+   * @param gse the {@link GeneralSecurityException} that needs to be set while invoking callbacks for the pending jobs
+   */
+  private void closePendingJobs(GeneralSecurityException gse) {
+    CryptoJob cryptoJob = jobQueue.poll();
+    while (cryptoJob != null) {
+      cryptoJob.closeJob(gse);
+      cryptoJob = jobQueue.poll();
+    }
+  }
+
+  /**
+   * CryptoJob representing the job that needs processing
+   */
+  static abstract class CryptoJob {
+    protected final BlobId blobId;
```

Comment:
I don't see a strong reason for making this class an abstract class. If you were doing a `getBlobId()` or `getInput()` on this class itself maybe there is a case. This is abstract just to reuse `blobId` and `input`. Otherwise this can simply be an interface, which is generally preferred.